flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/SpillingThread.java

```diff
@@ -0,0 +1,575 @@
+import org.apache.flink.runtime.io.disk.iomanager.BlockChannelReader;
+import org.apache.flink.runtime.io.disk.iomanager.BlockChannelWriter;
+import org.apache.flink.runtime.io.disk.iomanager.ChannelReaderInputView;
+import org.apache.flink.runtime.io.disk.iomanager.ChannelWriterOutputView;
+import org.apache.flink.runtime.io.disk.iomanager.FileIOChannel;
+import org.apache.flink.runtime.io.disk.iomanager.IOManager;
+import org.apache.flink.runtime.memory.MemoryManager;
+import org.apache.flink.runtime.util.EmptyMutableObjectIterator;
+import org.apache.flink.util.MutableObjectIterator;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Queue;
+
+import static org.apache.flink.runtime.operators.sort.CircularElement.EOF_MARKER;
+import static org.apache.flink.runtime.operators.sort.CircularElement.SPILLING_MARKER;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * The thread that handles the spilling of intermediate results and sets up the merging. It also merges the
+ * channels until sufficiently few channels remain to perform the final streamed merge.
+ */
+final class SpillingThread<E> extends ThreadBase<E> {
+	/**
+	 * An interface for injecting custom behaviour for spilling and merging phases.
+	 */
+	interface SpillingBehaviour<E> {
+		default void open() {}
+
+		default void close() {}
+
+		/**
+		 * A method that allows adjusting the spilling phase. We can inject e.g. combining the elements while spilling.
+		 */
+		void spillBuffer(
+			CircularElement<E> element,
+			ChannelWriterOutputView output,
+			LargeRecordHandler<E> largeRecordHandler) throws IOException;
+
+		/**
+		 * A method that allows adjusting the merging phase. We can inject e.g. combining the spilled elements.
+		 */
+		void mergeRecords(MergeIterator<E> mergeIterator, ChannelWriterOutputView output) throws IOException;
+	}
+
+	/** Logging. */
+	private static final Logger LOG = LoggerFactory.getLogger(SpillingThread.class);
+
+	private final MemoryManager memManager;			// memory manager to release memory
+
+	private final IOManager ioManager;				// I/O manager to create channels
+
+	private final TypeSerializer<E> serializer;		// The serializer for the data type
+
+	private final TypeComparator<E> comparator;		// The comparator that establishes the order relation.
+
+	private final List<MemorySegment> writeMemory;	// memory segments for writing
+
+	private final List<MemorySegment> mergeReadMemory;	// memory segments for sorting/reading
+
+	private final int maxFanIn;
+
+	private final SpillChannelManager spillChannelManager;
+
+	private final LargeRecordHandler<E> largeRecordHandler;
+
+	private final SpillingBehaviour<E> spillingBehaviour;
+
+	private boolean spillingBehaviourOpened = false;
```

Comment:
I think this should become `volatile` because it is called in `close()` and in `go()` which run on/ are called by different threads.

Reply:
Good call! I missed that.